### PR TITLE
Backtick delta table name when loading

### PR DESCRIPTION
**Pull request by annzhang-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/annzhang-db/mlflow/pull/13718?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13718/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13718
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
- Backtick table name when loading

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Before:
Screenshot 2024-11-08 at 3 23 57 PM

After:
Screenshot 2024-11-08 at 3 22 45 PM

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Review by harupy — APPROVED** [review]

**Review by shichengzhou-db — APPROVED** [review]

**Comment by shichengzhou-db** [comment]
thanks @annzhang-db!

File: mlflow/data/spark_delta_utils.py [hunk 1]
```diff
@@ -85,8 +85,9 @@ def _try_get_delta_table_latest_version_from_table_name(table_name: str) -> Opti
 
     try:
         spark = SparkSession.builder.getOrCreate()
+        backticked_table_name = ".".join(f"`{part}`" for part in table_name.split("."))
```
Comment (harupy) [inline mlflow/data/spark_delta_utils.py:88]: Is there any chance that `part` is already backticked?

Comment (harupy) [reply mlflow/data/spark_delta_utils.py:88]: /ai write `is_backticked` function in python.

File: mlflow/data/spark_delta_utils.py [hunk 2]
```diff
@@ -85,8 +87,11 @@ def _try_get_delta_table_latest_version_from_table_name(table_name: str) -> Opti
 
     try:
         spark = SparkSession.builder.getOrCreate()
+        backticked_table_name = ".".join(
+            part if _is_backticked(part) else f"`{part}`" for part in table_name.split(".")
+        )
```
Comment (serena-ruan) [inline mlflow/data/spark_delta_utils.py:92]: Can we make this a function?

Comment (harupy) [reply mlflow/data/spark_delta_utils.py:92]: I'd make that function if there were more names to backtick (say 5).

**Comment by harupy** [comment]
@annzhang-db @shichengzhou-db Do we really need this PR? The error message (`must be back quoted`) is clear and actionable.

File: mlflow/data/spark_delta_utils.py (hunk 2 above)
Comment (serena-ruan) [reply mlflow/data/spark_delta_utils.py:92]: But the same logic is used twice, I think `backtick_delta_table_name` is more helpful than `_is_backticked` method

Comment (harupy) [reply mlflow/data/spark_delta_utils.py:92]: ok, I'll make that funciton.

Comment (harupy) [reply mlflow/data/spark_delta_utils.py]: Here's the comparison of how I view this code:

```python
        # It's clear that each part of the table name should be backtick quoted
        backticked_table_name = ".".join(map(_backtick_quote, table_name.split(".")))
        # Does this quote the entire table name or each part?
        backticked_table_name = backtick_delta_table_name(table_name)
```

**Comment by shichengzhou-db** [comment]
> @harupy I believe the reason we wanted this PR is because when users pass in back quoted name, it saves the name incorrectly (with back quotes) which caused some lineage issue as this is inconsistent with the name in other places. @shichengzhou-db has more context on the issue.

Yes that's the gist of it.
To elaborate, we received an internal ES ticket(ES-1278130) where the customer trains a model with a UC table but no model -> table lineage information was recorded.

What happened was:
1) The customer passed in UC table name without backticks(e.g catalog.schema-name.table), pyspark complains about it, then customer tried passing catalog.`schema-name`.table. This is a valid name for pyspark, but not for UC. When mlflow attempts to retrieve id for the table, it fails silently and emits no lineage. Also, backticking seems to be a strange requirement from pyspark, we(as mlflow) want to only accept and store table name without any backticks.

**Pull request merged**